File: src/web-ui/src/infrastructure/api/service-api/MiniAppMarketAPI.ts
import { api } from './ApiClient';
import { createTauriCommandError } from '../errors/TauriCommandError';
import type {
  MiniApp,
  MiniAppI18n,
  MiniAppMeta,
  MiniAppPermissionDiff,
  MiniAppPermissions,
} from './MiniAppAPI';

export type MarketSort = 'newest' | 'downloads' | 'rating';
export type MarketSubmissionStatus =
  | 'draft'
  | 'submitted'
  | 'approved'
  | 'rejected'
  | 'withdrawn';

export interface MarketUserSummary {
  githubId: number;
  login: string;
  avatarUrl: string;
}

export interface MarketRelease {
  releaseId: string;
  listingId: string;
  releaseNumber: number;
  minBitfunVersion: string;
  changelog: string;
  packageSha256: string;
  packageSize: number;
  reviewBundleHash: string;
  permissions: MiniAppPermissions;
  publishedAt: number;
  yanked: boolean;
}

export interface MarketListingSummary {
  listingId: string;
  slug: string;
  name: string;
  description: string;
  icon: string;
  category: string;
  tags: string[];
  owner: MarketUserSummary;
  latestRelease: number;
  minBitfunVersion: string;
  permissions: MiniAppPermissions;
  screenshotUrls: string[];
  ratingAverage: number;
  ratingCount: number;
  favoriteCount: number;
  downloadCount: number;
  publishedAt: number;
  i18n?: MiniAppI18n;
  isFavorited?: boolean;
  myRating?: number;
}

export interface MarketListingDetail extends MarketListingSummary {
  changelog: string;
  license: { spdxExpression?: string; customUrl?: string };
  repositoryUrl?: string;
  releases: MarketRelease[];
}

export interface CursorPage<T> {
  items: T[];
  nextCursor?: string;
}

export interface MarketBrowseRequest {
  query?: string;
  category?: string;
  sort?: MarketSort;
  cursor?: string;
  limit?: number;
}

export interface MarketMe {
  user: MarketUserSummary;
  isAdmin: boolean;
}

export interface DesktopAuthStart {
  transactionId: string;
  transactionSecret: string;
  authorizationUrl: string;
  expiresAt: number;
  pollIntervalSeconds: number;
}

export interface MarketSubmission {
  submissionId: string;
  listingId?: string;
  slug: string;
  releaseNumber: number;
  name: string;
  description: string;
  icon: string;
  category: string;
  tags: string[];
  minBitfunVersion: string;
  changelog: string;
  license: { spdxExpression?: string; customUrl?: string };
  repositoryUrl?: string;
  permissions: MiniAppPermissions;
  status: MarketSubmissionStatus;
  packageSha256?: string;
  packageSize?: number;
  screenshotUrls: string[];
  rejectionReason?: string;
  createdAt: number;
  updatedAt: number;
}

export interface MarketSubmissionDraftRequest {
  listingId?: string;
  slug: string;
  releaseNumber: number;
  name: string;
  description: string;
  icon: string;
  category: string;
  tags: string[];
  minBitfunVersion: string;
  changelog: string;
  license: { spdxExpression?: string; customUrl?: string };
  repositoryUrl?: string;
}

export interface InstalledMarketOrigin {
  listingId: string;
  releaseId: string;
  releaseNumber: number;
  packageSha256: string;
}

export interface MarketInstalledStatus {
  appId: string;
  appVersion: number;
  permissions: MiniAppPermissions;
  origin: InstalledMarketOrigin;
  localOverride: boolean;
}

export interface MarketInstallResult {
  app: MiniApp;
  origin: InstalledMarketOrigin;
  updated: boolean;
  permissionDiff: MiniAppPermissionDiff;
}

export interface MarketPackageInspection {
  name: string;
  description: string;
  packageSha256: string;
  permissions: MiniAppPermissions;
  permissionDiff: MiniAppPermissionDiff;
}

export interface MarketUploadProgress {
  submissionId?: string;
  phase: 'validating' | 'package' | 'screenshots' | 'submitted';
  completed: number;
  total: number;
}

export class MiniAppMarketAPI {
  async browse(request: MarketBrowseRequest): Promise<CursorPage<MarketListingSummary>> {
    try {
      return await api.invoke('miniapp_market_browse', { request });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_browse', error);
    }
  }

  async getListing(slug: string): Promise<MarketListingDetail> {
    try {
      return await api.invoke('miniapp_market_get_listing', { request: { slug } });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_get_listing', error, { slug });
    }
  }

  async authStart(): Promise<DesktopAuthStart> {
    try {
      return await api.invoke('miniapp_market_auth_start', {});
    } catch (error) {
      throw createTauriCommandError('miniapp_market_auth_start', error);
    }
  }

  async authPoll(transaction: DesktopAuthStart): Promise<'pending' | 'authorized' | 'expired'> {
    try {
      const response = await api.invoke<{ status: 'pending' | 'authorized' | 'expired' }>(
        'miniapp_market_auth_poll',
        {
          request: {
            transactionId: transaction.transactionId,
            transactionSecret: transaction.transactionSecret,
          },
        },
      );
      return response.status;
    } catch (error) {
      throw createTauriCommandError('miniapp_market_auth_poll', error);
    }
  }

  async me(): Promise<MarketMe | null> {
    try {
      return await api.invoke('miniapp_market_me', {});
    } catch (error) {
      throw createTauriCommandError('miniapp_market_me', error);
    }
  }

  async logout(): Promise<void> {
    try {
      await api.invoke('miniapp_market_logout', {});
    } catch (error) {
      throw createTauriCommandError('miniapp_market_logout', error);
    }
  }

  async setRating(slug: string, value?: number): Promise<{
    average: number;
    count: number;
    myRating?: number;
  }> {
    try {
      return await api.invoke('miniapp_market_set_rating', {
        request: { slug, value },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_set_rating', error, { slug });
    }
  }

  async setFavorite(slug: string, enabled: boolean): Promise<{
    count: number;
    isFavorited: boolean;
  }> {
    try {
      return await api.invoke('miniapp_market_set_favorite', {
        request: { slug, enabled },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_set_favorite', error, { slug });
    }
  }

  async listSubmissions(): Promise<MarketSubmission[]> {
    try {
      return await api.invoke('miniapp_market_list_submissions', {});
    } catch (error) {
      throw createTauriCommandError('miniapp_market_list_submissions', error);
    }
  }

  async withdrawSubmission(submissionId: string): Promise<MarketSubmission> {
    try {
      return await api.invoke('miniapp_market_withdraw_submission', {
        request: { submissionId },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_withdraw_submission', error, {
        submissionId,
      });
    }
  }

  async installedStatus(listingId: string): Promise<MarketInstalledStatus | null> {
    try {
      return await api.invoke('miniapp_market_installed_status', { request: { listingId } });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_installed_status', error, { listingId });
    }
  }

  async install(
    slug: string,
    releaseNumber: number,
    options: {
      existingAppId?: string;
      confirmPermissions: boolean;
      confirmOverwrite: boolean;
    },
  ): Promise<MarketInstallResult> {
    try {
      return await api.invoke('miniapp_market_install', {
        request: {
          slug,
          releaseNumber,
          existingAppId: options.existingAppId,
          confirmPermissions: options.confirmPermissions,
          confirmOverwrite: options.confirmOverwrite,
        },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_install', error, { slug, releaseNumber });
    }
  }

  async inspectPackage(path: string): Promise<MarketPackageInspection> {
    try {
      return await api.invoke('miniapp_market_inspect_package', { request: { path } });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_inspect_package', error, { path });
    }
  }

  async captureWindow(): Promise<string> {
    try {
      return await api.invoke('miniapp_market_capture_window', {});
    } catch (error) {
      throw createTauriCommandError('miniapp_market_capture_window', error);
    }
  }

  async importPackage(path: string, confirmPermissions: boolean): Promise<MiniApp> {
    try {
      return await api.invoke('miniapp_market_import_package', {
        request: { path, confirmPermissions },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_import_package', error, { path });
    }
  }

  async submitInstalled(
    appId: string,
    draft: MarketSubmissionDraftRequest,
    screenshotPaths: string[],
  ): Promise<MarketSubmission> {
    try {
      return await api.invoke('miniapp_market_submit_installed', {
        request: { appId, draft, screenshotPaths },
      });
    } catch (error) {
      throw createTauriCommandError('miniapp_market_submit_installed', error, { appId });
    }
  }

  onUploadProgress(handler: (progress: MarketUploadProgress) => void): () => void {
    return api.listen<MarketUploadProgress>('miniapp-market-upload-progress', handler);
  }
}

export const miniAppMarketAPI = new MiniAppMarketAPI();

export type { MiniAppMeta };

File: src/apps/desktop/src/api/miniapp_market_api.rs
//! Native MiniApp marketplace adapter.
//!
//! Network and credential-vault details live in Services. This module maps the
//! shared marketplace DTOs into Tauri commands and owns the local install /
//! update transaction against the desktop MiniApp manager.

use crate::api::app_state::AppState;
use bitfun_core::miniapp::{
    MiniApp, MiniAppCustomizationMetadata, MiniAppPermissionDiff, MiniAppPermissions, MiniAppSource,
};
use bitfun_product_domains::miniapp::customization::diff_permissions;
use bitfun_product_domains::miniapp::market::{
    CursorPage, InstalledMarketOrigin, MarketListingDetail, MarketListingSummary, MarketRelease,
    MarketSubmission, MarketSubmissionDraftRequest,
};
use bitfun_services_integrations::miniapp_market::{
    submit_installed_app, validate_market_package, DesktopAuthPollRequest, DesktopAuthPollResponse,
    DesktopAuthStart, FavoriteAggregate, MarketBrowseRequest, MarketClient, MarketMe,
    RatingAggregate, ValidatedMarketPackage,
};
use serde::{Deserialize, Serialize};
use std::path::{Path, PathBuf};
use tauri::{AppHandle, Emitter, Manager, State, WebviewWindow};
use tokio::io::AsyncWriteExt;

const MARKET_UPLOAD_PROGRESS_EVENT: &str = "miniapp-market-upload-progress";

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketInstallRequest {
    pub slug: String,
    pub release_number: u32,
    pub existing_app_id: Option<String>,
    #[serde(default)]
    pub confirm_permissions: bool,
    #[serde(default)]
    pub confirm_overwrite: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketInstallResult {
    pub app: MiniApp,
    pub origin: InstalledMarketOrigin,
    pub updated: bool,
    pub permission_diff: MiniAppPermissionDiff,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketInstalledStatus {
    pub app_id: String,
    pub app_version: u32,
    pub permissions: MiniAppPermissions,
    pub origin: InstalledMarketOrigin,
    pub local_override: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketSetRatingRequest {
    pub slug: String,
    pub value: Option<u8>,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketSetFavoriteRequest {
    pub slug: String,
    pub enabled: bool,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketSlugRequest {
    pub slug: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketSubmissionIdRequest {
    pub submission_id: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketListingIdRequest {
    pub listing_id: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MarketPackagePathRequest {
    pub path: String,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketSubmitInstalledRequest {
    pub app_id: String,
    pub draft: MarketSubmissionDraftRequest,
    pub screenshot_paths: Vec<String>,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
struct MarketUploadProgress {
    submission_id: Option<String>,
    phase: &'static str,
    completed: u32,
    total: u32,
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketImportPackageRequest {
    pub path: String,
    #[serde(default)]
    pub confirm_permissions: bool,
}

#[derive(Debug, Clone, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct MarketPackageInspection {
    pub name: String,
    pub description: String,
    pub package_sha256: String,
    pub permissions: MiniAppPermissions,
    pub permission_diff: MiniAppPermissionDiff,
}

#[tauri::command]
pub async fn miniapp_market_browse(
    request: MarketBrowseRequest,
) -> Result<CursorPage<MarketListingSummary>, String> {
    let client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.browse(&request).await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_get_listing(
    request: MarketSlugRequest,
) -> Result<MarketListingDetail, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.listing(&request.slug).await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_auth_start() -> Result<DesktopAuthStart, String> {
    let client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.start_desktop_auth().await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_auth_poll(
    request: DesktopAuthPollRequest,
) -> Result<DesktopAuthPollResponse, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client
        .poll_desktop_auth(&request)
        .await
        .map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_me() -> Result<Option<MarketMe>, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.me().await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_logout() -> Result<(), String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.logout().await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_set_rating(
    request: MarketSetRatingRequest,
) -> Result<RatingAggregate, String> {
    if request.value.is_some_and(|value| !(1..=5).contains(&value)) {
        return Err("Rating must be between 1 and 5.".to_string());
    }
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client
        .set_rating(&request.slug, request.value)
        .await
        .map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_set_favorite(
    request: MarketSetFavoriteRequest,
) -> Result<FavoriteAggregate, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client
        .set_favorite(&request.slug, request.enabled)
        .await
        .map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_list_submissions() -> Result<Vec<MarketSubmission>, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client.list_submissions().await.map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_withdraw_submission(
    request: MarketSubmissionIdRequest,
) -> Result<MarketSubmission, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    client
        .withdraw_submission(&request.submission_id)
        .await
        .map_err(market_error)
}

#[tauri::command]
pub async fn miniapp_market_installed_status(
    state: State<'_, AppState>,
    request: MarketListingIdRequest,
) -> Result<Option<MarketInstalledStatus>, String> {
    let apps = state
        .miniapp_manager
        .list()
        .await
        .map_err(|error| error.to_string())?;
    for app in apps {
        let Some(metadata) = state
            .miniapp_manager
            .load_customization_metadata(&app.id)
            .await
            .map_err(|error| error.to_string())?
        else {
            continue;
        };
        let Some(origin) = metadata.origin.market else {
            continue;
        };
        if origin.listing_id == request.listing_id {
            return Ok(Some(MarketInstalledStatus {
                app_id: app.id,
                app_version: app.version,
                permissions: app.permissions,
                origin,
                local_override: metadata.local_override,
            }));
        }
    }
    Ok(None)
}

#[tauri::command]
pub async fn miniapp_market_install(
    state: State<'_, AppState>,
    request: MarketInstallRequest,
) -> Result<MarketInstallResult, String> {
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    let detail = client.listing(&request.slug).await.map_err(market_error)?;
    let release = find_release(&detail, request.release_number)?;
    validate_minimum_bitfun_version(&release.min_bitfun_version)?;
    if release.yanked {
        return Err(
            "This marketplace release has been yanked and cannot be installed.".to_string(),
        );
    }

    let bytes = client
        .download_release(&detail.summary.slug, release.release_number)
        .await
        .map_err(market_error)?;
    let package = stage_downloaded_package(&bytes).await?;
    if package.sha256 != release.package_sha256 || package.size != release.package_size {
        return Err("The downloaded package does not match the reviewed release.".to_string());
    }
    if package.meta.permissions != release.permissions {
        return Err("The package permissions do not match the reviewed release.".to_string());
    }
    let origin = InstalledMarketOrigin {
        listing_id: release.listing_id.clone(),
        release_id: release.release_id.clone(),
        release_number: release.release_number,
        package_sha256: release.package_sha256.clone(),
    };
    let source = source_from_package(&package);

    let (app, updated, permission_diff) = if let Some(app_id) = request.existing_app_id {
        let previous = state
            .miniapp_manager
            .get(&app_id)
            .await
            .map_err(|error| error.to_string())?;
        let metadata = require_matching_market_origin(
            state
                .miniapp_manager
                .load_customization_metadata(&app_id)
                .await
                .map_err(|error| error.to_string())?,
            &origin.listing_id,
        )?;
        if metadata.local_override && !request.confirm_overwrite {
            return Err(
                "This MiniApp has local customizations. Confirm overwrite to create a snapshot and continue."
                    .to_string(),
            );
        }
        let diff = diff_permissions(&previous.permissions, &package.meta.permissions);
        require_permission_confirmation(&diff, request.confirm_permissions)?;
        let app = state
            .miniapp_manager
            .update_market_package(&app_id, package.meta, source, origin.clone())
            .await
            .map_err(|error| error.to_string())?;
        (app, true, diff)
    } else {
        let diff = diff_permissions(&MiniAppPermissions::default(), &package.meta.permissions);
        require_permission_confirmation(&diff, request.confirm_permissions)?;
        let app = state
            .miniapp_manager
            .install_market_package(package.meta, source, origin.clone())
            .await
            .map_err(|error| error.to_string())?;
        (app, false, diff)
    };
    Ok(MarketInstallResult {
        app,
        origin,
        updated,
        permission_diff,
    })
}

async fn stage_downloaded_package(bytes: &[u8]) -> Result<ValidatedMarketPackage, String> {
    if bytes.len() as u64 > bitfun_product_domains::miniapp::market::MARKET_MAX_PACKAGE_BYTES {
        return Err("The downloaded package exceeds 20 MiB.".to_string());
    }
    let directory = tempfile::Builder::new()
        .prefix("bitfun-miniapp-market-download-")
        .tempdir()
        .map_err(|error| format!("Could not create a private download directory: {error}"))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        tokio::fs::set_permissions(directory.path(), std::fs::Permissions::from_mode(0o700))
            .await
            .map_err(|error| format!("Could not secure the download directory: {error}"))?;
    }
    let path = directory.path().join("package.bfminiapp");
    let mut file = tokio::fs::OpenOptions::new()
        .write(true)
        .create_new(true)
        .open(&path)
        .await
        .map_err(|error| format!("Could not stage the downloaded package: {error}"))?;
    file.write_all(bytes)
        .await
        .map_err(|error| format!("Could not write the downloaded package: {error}"))?;
    file.sync_all()
        .await
        .map_err(|error| format!("Could not flush the downloaded package: {error}"))?;
    drop(file);
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        tokio::fs::set_permissions(&path, std::fs::Permissions::from_mode(0o600))
            .await
            .map_err(|error| format!("Could not secure the downloaded package: {error}"))?;
    }
    read_and_validate_package_file(&path).await
}

#[tauri::command]
pub async fn miniapp_market_import_package(
    state: State<'_, AppState>,
    request: MarketImportPackageRequest,
) -> Result<MiniApp, String> {
    let path = PathBuf::from(&request.path);
    let metadata = tokio::fs::metadata(&path)
        .await
        .map_err(|error| format!("Could not read package metadata: {error}"))?;
    if !metadata.is_file()
        || metadata.len() > bitfun_product_domains::miniapp::market::MARKET_MAX_PACKAGE_BYTES
    {
        return Err("The selected .bfminiapp file is invalid or exceeds 20 MiB.".to_string());
    }
    if path
        .extension()
        .and_then(|extension| extension.to_str())
        .is_none_or(|extension| !extension.eq_ignore_ascii_case("bfminiapp"))
    {
        return Err("Select a .bfminiapp package.".to_string());
    }
    let bytes = tokio::fs::read(&path)
        .await
        .map_err(|error| format!("Could not read package: {error}"))?;
    let package = validate_market_package(&bytes).map_err(|error| error.to_string())?;
    let diff = diff_permissions(&MiniAppPermissions::default(), &package.meta.permissions);
    require_permission_confirmation(&diff, request.confirm_permissions)?;
    let source = source_from_package(&package);
    state
        .miniapp_manager
        .install_strict_package_import(package.meta, source)
        .await
        .map_err(|error| error.to_string())
}

#[tauri::command]
pub async fn miniapp_market_inspect_package(
    request: MarketPackagePathRequest,
) -> Result<MarketPackageInspection, String> {
    let package = read_and_validate_package_file(Path::new(&request.path)).await?;
    let permission_diff =
        diff_permissions(&MiniAppPermissions::default(), &package.meta.permissions);
    Ok(MarketPackageInspection {
        name: package.meta.name.clone(),
        description: package.meta.description.clone(),
        package_sha256: package.sha256.clone(),
        permissions: package.meta.permissions,
        permission_diff,
    })
}

#[tauri::command]
pub async fn miniapp_market_capture_window(
    app: AppHandle,
    window: WebviewWindow,
) -> Result<String, String> {
    let position = window
        .outer_position()
        .map_err(|error| format!("Could not read the BitFun window position: {error}"))?;
    let size = window
        .outer_size()
        .map_err(|error| format!("Could not read the BitFun window size: {error}"))?;
    if size.width < 320 || size.height < 240 {
        return Err("The BitFun window is too small to capture a review screenshot.".to_string());
    }

    let capture_dir = app
        .path()
        .app_cache_dir()
        .map_err(|error| format!("Could not resolve the private cache directory: {error}"))?
        .join("miniapp-market-captures");
    tokio::fs::create_dir_all(&capture_dir)
        .await
        .map_err(|error| format!("Could not create the screenshot directory: {error}"))?;
    #[cfg(unix)]
    {
        use std::os::unix::fs::PermissionsExt;
        tokio::fs::set_permissions(&capture_dir, std::fs::Permissions::from_mode(0o700))
            .await
            .map_err(|error| format!("Could not secure the screenshot directory: {error}"))?;
    }
    let path = capture_dir.join(format!("{}.jpg", uuid::Uuid::new_v4()));
    let output_path = path.clone();

    tokio::task::spawn_blocking(move || -> Result<(), String> {
        let center_x = position.x.saturating_add((size.width / 2) as i32);
        let center_y = position.y.saturating_add((size.height / 2) as i32);
        let screen = screenshots::Screen::from_point(center_x, center_y)
            .map_err(|error| format!("Could not access the current display: {error}"))?;
        let relative_x = position.x.saturating_sub(screen.display_info.x);
        let relative_y = position.y.saturating_sub(screen.display_info.y);
        let captured = screen
            .capture_area(relative_x, relative_y, size.width, size.height)
            .map_err(|error| {
                format!(
                    "Window capture failed. On macOS, grant BitFun Screen Recording permission: {error}"
                )
            })?;
        let (width, height) = captured.dimensions();
        let rgba = image::RgbaImage::from_raw(width, height, captured.into_raw())
            .ok_or_else(|| "The captured window had an invalid pixel buffer.".to_string())?;
        let image = image::DynamicImage::ImageRgba8(rgba);
        let resized = if width > 1600 || height > 1200 {
            image.resize(1600, 1200, image::imageops::FilterType::Lanczos3)
        } else {
            image
        };
        let file = std::fs::File::create(&output_path)
            .map_err(|error| format!("Could not create the screenshot: {error}"))?;
        let mut encoder = image::codecs::jpeg::JpegEncoder::new_with_quality(file, 88);
        encoder
            .encode_image(&resized)
            .map_err(|error| format!("Could not encode the screenshot: {error}"))
    })
    .await
    .map_err(|error| format!("Screenshot capture task failed: {error}"))??;

    Ok(path.to_string_lossy().into_owned())
}

#[tauri::command]
pub async fn miniapp_market_submit_installed(
    app: AppHandle,
    state: State<'_, AppState>,
    request: MarketSubmitInstalledRequest,
) -> Result<MarketSubmission, String> {
    let installed = state
        .miniapp_manager
        .get(&request.app_id)
        .await
        .map_err(|error| error.to_string())?;
    let mut client = MarketClient::from_environment()
        .await
        .map_err(market_error)?;
    let mut progress = |submission_id: Option<&str>, phase: &'static str, completed, total| {
        emit_upload_progress(&app, submission_id, phase, completed, total);
    };
    let submission = submit_installed_app(
        &mut client,
        &installed,
        &request.draft,
        &request.screenshot_paths,
        &mut progress,
    )
    .await
    .map_err(market_error)?;
    remove_owned_capture_files(&app, &request.screenshot_paths).await;
    Ok(submission)
}

async fn remove_owned_capture_files(app: &AppHandle, paths: &[String]) {
    let Ok(capture_dir) = app
        .path()
        .app_cache_dir()
        .map(|path| path.join("miniapp-market-captures"))
    else {
        return;
    };
    for value in paths {
        let path = PathBuf::from(value);
        if path.parent() == Some(capture_dir.as_path()) {
            if let Err(error) = tokio::fs::remove_file(&path).await {
                if error.kind() != std::io::ErrorKind::NotFound {
                    log::warn!(
                        "remove submitted MiniApp market capture {} failed: {}",
                        path.display(),
                        error
                    );
                }
            }
        }
    }
}

fn source_from_package(package: &ValidatedMarketPackage) -> MiniAppSource {
    let file = |name: &str| package.source_files.get(name).cloned().unwrap_or_default();
    MiniAppSource {
        html: file("source/index.html"),
        css: file("source/style.css"),
        ui_js: file("source/ui.js"),
        esm_dependencies: Vec::new(),
        worker_js: file("source/worker.js"),
        npm_dependencies: Vec::new(),
    }
}

async fn read_and_validate_package_file(path: &Path) -> Result<ValidatedMarketPackage, String> {
    let metadata = tokio::fs::metadata(path)
        .await
        .map_err(|error| format!("Could not read package metadata: {error}"))?;
    if !metadata.is_file()
        || metadata.len() > bitfun_product_domains::miniapp::market::MARKET_MAX_PACKAGE_BYTES
    {
        return Err("The selected .bfminiapp file is invalid or exceeds 20 MiB.".to_string());
    }
    if path
        .extension()
        .and_then(|extension| extension.to_str())
        .is_none_or(|extension| !extension.eq_ignore_ascii_case("bfminiapp"))
    {
        return Err("Select a .bfminiapp package.".to_string());
    }
    let bytes = tokio::fs::read(path)
        .await
        .map_err(|error| format!("Could not read package: {error}"))?;
    validate_market_package(&bytes).map_err(|error| error.to_string())
}

fn find_release(
    detail: &MarketListingDetail,
    release_number: u32,
) -> Result<MarketRelease, String> {
    detail
        .releases
        .iter()
        .find(|release| release.release_number == release_number)
        .cloned()
        .ok_or_else(|| format!("Marketplace release {release_number} was not found."))
}

fn require_matching_market_origin(
    metadata: Option<MiniAppCustomizationMetadata>,
    listing_id: &str,
) -> Result<MiniAppCustomizationMetadata, String> {
    let metadata = metadata.ok_or_else(|| "MiniApp has no marketplace origin.".to_string())?;
    if metadata
        .origin
        .market
        .as_ref()
        .is_none_or(|origin| origin.listing_id != listing_id)
    {
        return Err(
            "The installed MiniApp belongs to a different marketplace listing.".to_string(),
        );
    }
    Ok(metadata)
}

fn require_permission_confirmation(
    diff: &MiniAppPermissionDiff,
    confirmed: bool,
) -> Result<(), String> {
    if (!diff.added.is_empty() || !diff.expanded.is_empty()) && !confirmed {
        return Err(format!(
            "Permission confirmation required: {}",
            diff.added
                .iter()
                .chain(diff.expanded.iter())
                .cloned()
                .collect::<Vec<_>>()
                .join(", ")
        ));
    }
    Ok(())
}

fn validate_minimum_bitfun_version(minimum: &str) -> Result<(), String> {
    let minimum = semver::Version::parse(minimum)
        .map_err(|_| "The release declares an invalid minimum BitFun version.".to_string())?;
    let current = semver::Version::parse(env!("CARGO_PKG_VERSION"))
        .map_err(|_| "The current BitFun version is invalid.".to_string())?;
    if current < minimum {
        return Err(format!(
            "This MiniApp requires BitFun {minimum} or newer. Current version: {current}."
        ));
    }
    Ok(())
}

fn emit_upload_progress(
    app: &AppHandle,
    submission_id: Option<&str>,
    phase: &'static str,
    completed: u32,
    total: u32,
) {
    let _ = app.emit(
        MARKET_UPLOAD_PROGRESS_EVENT,
        MarketUploadProgress {
            submission_id: submission_id.map(str::to_string),
            phase,
            completed,
            total,
        },
    );
}

fn market_error(error: impl Serialize + std::fmt::Display) -> String {
    serde_json::to_string(&error).unwrap_or_else(|_| error.to_string())
}
